optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java

```diff
@@ -192,6 +195,42 @@
                 resultContainer -> resultContainer[0]);
     }
 
+    // ************************************************************************
+    // min
+    // ************************************************************************
+
+    private static <A> UniConstraintCollector<A, SortedMap<A, Long>, A> minOrMax(Comparator<A> comparator,
+            boolean min) {
+        Function<SortedMap<A, Long>, A> keySupplier = min ? SortedMap::firstKey : SortedMap::lastKey;
+        return new DefaultUniConstraintCollector<>(
+                () -> new TreeMap<>(comparator),
+                (resultContainer, a) -> {
+                    resultContainer.compute(a, (key, value) -> value == null ? 1 : value + 1);
```

Comment:
This leaves null entries in the map, which is effectively a memory leak.

Example:
Map starts out as {}
insert "Apple": {"Apple":1}
insert "Orange": {"Apple":1, "Orange":1}
insert "Apple": {"Apple":2, "Orange":1}
remove "Orange": {"Apple":2, "Orange":null}
...
A problem fact change might have removed the Orange, but it will linger in memory forever.

Reply:
I don't think that this is the case. If it were, the `finisher()` would not work. It checks for `size() == 0`, and the tests make sure that this is, in fact, the case.

Therefore I am assuming that returning `null` from `map.compute()` does, in fact, remove the entry. But I will play around with the `Map` to verify the assumption.

Reply:
You're right, no need to check:
https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#compute-K-java.util.function.BiFunction-
"If the function returns null, the mapping is removed (or remains absent if initially absent)."

Reply:
I suspect there is more code in optaplanner that does map.remove or removeKey but could benefit from this.